cypress-io/cypress
# feat: Support dependencies in cy.origin() callback

- Closes #19974

### User facing changelog

- Added support for requiring dependencies within the `cy.origin()` callback. See [insert docs link here] for more information.

### Additional details

Enables support for requiring dependencies within the `cy.origin()` callback using the `Cypress.require()` method.

```js
cy.visit('/fixtures/primary-origin.html')
cy.get('a[data-cy="cross-origin-secondary-link"]').click()
cy.origin('http://foobar.com:3500', () => {
  const _ = Cypress.require('lodash')

  expect(_.get({ foo: 'foo' }, 'foo')).to.equal('foo')
})
```

It's implemented by adding a loader to the webpack-preprocessor that extracts the callback when it contains a reference to `Cypress.require()`. It replaces the callback with an object that is used at runtime to reference and call into the extracted callback. It replaces any `Cypress.require()` uses with standard `require()`s and then runs the callback through webpack again so that it has its own bundle with its dependencies. 

One caveat of this is that the latest webpack preprocessor is required for this functionality. If it's not used or if an older version without the loader is used, `Cypress.require()` will not be replaced. Its default functionality is to error with a message that the latest webpack preprocessor is required.

### Steps to test

- Run Cypress with `experimentalSessionAndOrigin: true`
- Run a test like the following (this one will work in the driver):

```js
cy.visit('/fixtures/primary-origin.html')
cy.get('a[data-cy="cross-origin-secondary-link"]').click()
cy.origin('http://foobar.com:3500', () => {
  const _ = Cypress.require('lodash')

  expect(_.get({ foo: 'foo' }, 'foo')).to.equal('foo')
})
```

### How has the user experience changed?

Users can now use 1st and 3rd party dependencies in the `cy.origin()` callback and more easily share common code and custom commands between cross-origin tests.

### PR Tasks


- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/4681
